node/silkworm/db/buffer.cpp

```diff
@@ -132,21 +131,17 @@ void Buffer::write_to_state_table() {
         }
 
         if (auto it{storage_.find(address)}; it != storage_.end()) {
-            for (const auto& contract : it->second) {
-                uint64_t incarnation{contract.first};
+            for (const auto& [incarnation, contract_storage] : it->second) {
                 Bytes prefix{storage_prefix(address, incarnation)};
 
-                const auto& contract_storage{contract.second};
-
                 // sort before inserting into the DB
-                storage_keys.clear();
-                for (const auto& x : contract_storage) {
-                    storage_keys.push_back(x.first);
+                std::vector<evmc::bytes32> locations;
```

Comment:
This is probably a pessimization because you allocate and deallocate memory for `locations` for each loop iteration instead of growing capacity were it declared outside the loop.

Reply:
Yes you're right I forgot to keep the outside loop declaration and use clear on each iteration.
But thinking more I believe the sorting of locations is completely unneeded : we're not appending to table but using `upsert_storage_value` which makes use of `find_value_suffix(state_cursor, storage_prefix, location)` to point to the exact dup'ed record (if any) and we upsert into table.
Hence I beleive the code 
```
                // sort before inserting into the DB
                std::vector<evmc::bytes32> locations;
                for (const auto& [location, _] : contract_storage) {
                    locations.push_back(location);
                }
                std::sort(locations.begin(), locations.end());
                for (const auto& location : locations) {
                    upsert_storage_value(state_table, prefix, location, contract_storage.at(location));
                }
```
should be simplified to 
```
                Bytes prefix{storage_prefix(address, incarnation)};
                for (const auto& [location, value] : contract_storage) {
                    upsert_storage_value(state_table, prefix, location, value);
                }
```
which saves a copy, a sort, a double lookup in contract_storage map

Reply:
I'm also not sure processing addresses in a sorted way brings any benefit.
```
    // Extract sorted index of unique addresses before inserting into the DB
    absl::btree_set<evmc::address> addresses;
    for (auto& x : accounts_) {
        addresses.insert(x.first);
    }
    for (auto& x : storage_) {
        addresses.insert(x.first);
    }
```
This adds overhead and forces lookups on both `accounts_` and `storage_` maps relying on the assumption "adjacent addresses" can be adjacent also on db which is not true.

Reply:
It _might_ bring some benefit if we allocate the `storage_prefix` key outside the loop and do a `memcpy` only when address changes (instead of building the key for each record in `storage_` map) ... but I believe storage changes are way less than EOA Accounts changes per block

Reply:
Also noticed `block_account_changes_` records also changes for identical accounts. 
See https://github.com/torquem-ch/silkworm/blob/master/node/silkworm/db/buffer.cpp#L54-L63

Reply:
> It _might_ bring some benefit if we allocate the `storage_prefix` key outside the loop and do a `memcpy` only when address changes (instead of building the key for each record in `storage_` map) ... but I believe storage changes are way less than EOA Accounts changes per block

That's why we have to measure and it'd better be in separate PR. But anyway, I won't insist.

Reply:
And what about this ? https://github.com/torquem-ch/silkworm/pull/553#discussion_r798842253

Reply:
Again, we have to measure.